### PR TITLE
Add THREE.AudioLoader

### PR DESCRIPTION
**Pull request by EvilPeanut**
I have added `THREE.AudioLoader` to asynchronously load an `AudioBuffer` and provide callbacks for `onLoad`, `onProgress` and `onError`. Related issues: #8176, #8367.

The `THREE.AudioLoader` depreciates `THREE.AudioBuffer` and `THREE.Audio`'s `.load`, I have added depreciation messages to the old functions.

I have also made a documentation page for `THREE.AudioLoader` which is included.

Here is an example how `THREE.AudioLoader` can be used to load and play a sound once loaded

```
// instantiate a listener
var audioListener = new THREE.AudioListener();

// add the listener to the camera
camera.add( audioListener );

// instantiate audio object
var oceanAmbientSound = new THREE.Audio( audioListener );

// add the audio object to the scene
scene.add( oceanAmbientSound );

// instantiate a loader
var loader = new THREE.AudioLoader();

// load a resource
loader.load(
    // resource URL
    'audio/ambient_ocean.ogg',
    // Function when resource is loaded
    function ( audioBuffer ) {
        // set the audio object buffer to the loaded object
        oceanAmbientSound.setBuffer( audioBuffer );

        // play the audio
        oceanAmbientSound.play();
    },
    // Function called when download progresses
    function ( xhr ) {
        console.log( (xhr.loaded / xhr.total * 100) + '% loaded' );
    },
    // Function called when download errors
    function ( xhr ) {
        console.log( 'An error happened' );
    }
);
```

### CODE REVIEW

**Comment by Mugen87** [comment]
Hi Reece! Can you also adjust the following [sound showcase](http://threejs.org/examples/#misc_sound) in the three.js example directory? I'm afraid your changes will break the current implementation.

Example:
```javascript
var sound1 = new THREE.PositionalAudio( listener );
sound1.load( 'sounds/358232_j_s_song.ogg' ); // this method would be deprecated
```

Besides, it's maybe a good idea to place the deprecated code into `Three.Legacy.js`. The actual source files are more clear then.

**Comment by Mugen87** [comment]
Do you think there is a way to use `THREE.AudioLoader` in `THREE.AudioBuffer .load` to provide backwards compatibility? It would be good if `THREE.AudioBuffer` and `THREE.Audio .load` are still working after your changes (even when they deprecated).

**Comment by EvilPeanut** [comment]
@Mugen87 Thanks for your suggestions :) I will update the example now and add `THREE.AudioLoader` in the deprecated functions

**Comment by EvilPeanut** [comment]
@Mugen87 My latest commits should tackle your suggestions

**Comment by Mugen87** [comment]
Looks good! After studying at your changes, i realize that the code in `THREE.AudioBuffer .load` is not necessary after all. So i think we can remove it and just log the warning. Before doing that, we should consider some other opinions :wink: 

@mrdoob What do you think about @EvilPeanut's implementation? Using `THREE.AudioLoader` in this way would make `THREE.AudioBuffer` obsolet. I assume `THREE.AudioBuffer` is more an internal entity of `THREE.Audio` and not used by users of the library. Do you think we can remove it?

**Comment by Mugen87** [comment]
@EvilPeanut One more thing :wink:. Can you add `THREE.AudioLoader` to `utils/build/includes/common.json` (see [here](https://github.com/mrdoob/three.js/blob/master/utils/build/includes/common.json#L85))? If you don't do this, `THREE.AudioLoader` won't be part of the build.

**Comment by EvilPeanut** [comment]
@Mugen87 Added it :)

File: src/audio/Audio.js [hunk 1]
```diff
@@ -58,23 +62,19 @@ THREE.Audio.prototype.setNodeSource = function ( audioNode ) {
...
-	} );
+	scope.source.buffer = audioBuffer;
+	scope.sourceType = 'buffer';
+	if ( scope.autoplay ) scope.play();
 
 	return this;
 
 };
 
-THREE.Audio.prototype.play = function () {
```
Comment (mrdoob) [inline src/audio/Audio.js]: Seems like you removed the space in `function () {` in a bunch of places. Do you mind reverting that?

**Comment by mrdoob** [comment]
> @mrdoob What do you think about @EvilPeanut's implementation? Using `THREE.AudioLoader` in this way would make `THREE.AudioBuffer` obsolet. I assume `THREE.AudioBuffer` is more an internal entity of `THREE.Audio` and not used by users of the library. Do you think we can remove it?

Sounds good to me 😊

File: src/audio/Audio.js (hunk 1 above)
Comment (EvilPeanut) [inline src/audio/Audio.js]: I'll change that now. Sorry for these little mistakes. It is my first time contributing :).

I have tried to follow the contribution guidelines though and used http://zz85.github.io/mrdoobapproves/ which is why I removed the spaces, is it an issue with the site?

image

**Comment by EvilPeanut** [comment]
@mrdoob @Mugen87 I have removed `THREE.AudioBuffer` and added back the removed whitespace :smiley:

File: src/audio/Audio.js (hunk 1 above)
Comment (mrdoob) [inline src/audio/Audio.js]: It is an issue with the site indeed.
@zz85 do you mind addressing this?

**Comment by mrdoob** [comment]
Looks good!

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
FYI I've added `THREE.AudioContext`. That way we don't have pass the audio context around... https://github.com/mrdoob/three.js/commit/e207eb57ca8e6838b12a73db70486e3315647960

**Pull request merged**